From c38909f338b855b01c8f9694abc47200ab2d7d2b Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:12 -0400
Subject: [ethernet] mlx4: Fix memory leaks in SR-IOV error paths

Message-id: <d3de0fc31420f356814bdee061f431616273a3a5.1410425016.git.avadai@redhat.com>
Patchwork-id: 93661
O-Subject: [Patch RHEL7.1 V1 136/197] mlx4_core: Fix memory leaks in SR-IOV error paths
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Dotan Barak <dotanb@dev.mellanox.co.il>

Fix a few memory leaks that happen if errors happen in SR-IOV mode.

Signed-off-by: Dotan Barak <dotanb@dev.mellanox.co.il>
Signed-off-by: Jack Morgenstein <jackm@dev.mellanox.co.il>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit b38f2879b7d2d51747de2ea9062c698a6ac64cb1)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/main.c b/drivers/net/ethernet/mellanox/mlx4/main.c
index cf5faaf..ad154bb 100644
--- a/drivers/net/ethernet/mellanox/mlx4/main.c
+++ b/drivers/net/ethernet/mellanox/mlx4/main.c
@@ -1678,6 +1678,13 @@ unmap_bf:
  unmap_internal_clock(dev);
  unmap_bf_area(dev);
 
+ if (mlx4_is_slave(dev)) {
+  kfree(dev->caps.qp0_tunnel);
+  kfree(dev->caps.qp0_proxy);
+  kfree(dev->caps.qp1_tunnel);
+  kfree(dev->caps.qp1_proxy);
+ }
+
 err_close:
  if (mlx4_is_slave(dev))
   mlx4_slave_exit(dev);
@@ -2522,6 +2529,13 @@ err_master_mfunc:
  if (mlx4_is_master(dev))
   mlx4_multi_func_cleanup(dev);
 
+ if (mlx4_is_slave(dev)) {
+  kfree(dev->caps.qp0_tunnel);
+  kfree(dev->caps.qp0_proxy);
+  kfree(dev->caps.qp1_tunnel);
+  kfree(dev->caps.qp1_proxy);
+ }
+
 err_close:
  if (dev->flags & MLX4_FLAG_MSI_X)
   pci_disable_msix(pdev);
-- 
1.7.1